src/TestResult.dfy

```diff
@@ -49,4 +51,93 @@ module TestResult {
     var resourceCount :- Externs.ParseNat(resourceCountStr);
     Success(TestResult(displayName, outcome, durationTicks, resourceCount))
   }
-}
\ No newline at end of file
+
+  method GroupTestResults(results: seq<TestResult>) returns (groupedResults: map<string, seq<TestResult>>) {
+    groupedResults := map[];
+    for resIndex := 0 to |results| {
+      var res := results[resIndex];
+      var name := res.displayName;
+      if name in groupedResults {
+        groupedResults := groupedResults[name := groupedResults[name] + [res]];
+      } else {
+        groupedResults := groupedResults[name := [res]];
+      }
+    }
+  }
+
+  function method TestResultStatistics(results: seq<TestResult>, f: TestResult -> int): Statistics.Statistics
+  {
+    match 0 < |results| {
+      case true =>
+        var min := Seq.Min(Seq.Map(f, results));
+        var max := Seq.Max(Seq.Map(f, results));
+        var mean := Statistics.Mean(Seq.Map(x => f(x) as real, results));
+        var stddev := Statistics.StdDev(Seq.Map(x => f(x) as real, results));
+        Statistics.Statistics(min as real, max as real, mean, stddev)
+      case false =>
+        Statistics.Statistics(0.0, 0.0, 0.0, 0.0)
+    }
+  }
+
+  function method TestResultDurationStatistics(results: seq<TestResult>): Statistics.Statistics
+  {
+    TestResultStatistics(results, (result: TestResult) => result.durationTicks as int)
+  }
+
+  function method TestResultResourceStatistics(results: seq<TestResult>): Statistics.Statistics
+  {
+    TestResultStatistics(results, (result: TestResult) => result.resourceCount)
+  }
+
+  method PrintTestResultStatistics(displayName: string, results: seq<TestResult>)
+  {
+    var timeStats := TestResultStatistics(results, (r: TestResult) => r.durationTicks as int);
+    var resStats := TestResultStatistics(results, (r: TestResult) => r.resourceCount as int);
+    print displayName, "\n";
+    print "  Time - ", Statistics.StatisticsToSeconds(timeStats).ToString(), "\n";
+    print "  Resource Count - ", resStats.ToString(), "\n";
+  }
+
+  method PrintAllTestResultStatistics(groupedResults: map<string, seq<TestResult>>)
+  {
+    var resultBatches := groupedResults.Items;
+    while resultBatches != {}
+      decreases |resultBatches|
+    {
+      var resultBatch :| resultBatch in resultBatches;
+      resultBatches := resultBatches - {resultBatch};
+      if 0 < |resultBatch.1| { // TODO: could prove that this never happens
```

Comment:
It would be nice to prove, but I'm not yet sure if it's worth the effort. We'd need to track that we have a map full of sequences that are all non-empty.

Reply:
What about this:
```
type NonEmptySeq<T(0)> = x : seq<T> | |x| > 0 witness *
...

method PrintAllTestResultStatistics(groupedResults: map<string, NonEmptySeq<TestResult>>)

....
method GroupTestResults(results: seq<TestResult>) returns (groupedResults: map<string, NonEmptySeq<TestResult>>) {
```
No need to keep track of invariants, subsets type can do it for you.
I wish we could write an explicit witness but I don't know how.